crates/sui-framework-snapshot/tests/compatibility_tests.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+mod compatibility_tests {
+    use std::collections::BTreeMap;
+    use sui_framework::{compare_system_package, BuiltInFramework};
+    use sui_framework_snapshot::{load_bytecode_snapshot, load_bytecode_snapshot_manifest};
+    use sui_protocol_config::{ProtocolConfig, ProtocolVersion};
+
+    #[tokio::test]
+    async fn test_framework_compatibility() {
+        // This test checks that the current framework is compatible with all previous framework
+        // bytecode snapshots.
+        for (network, snapshots) in load_bytecode_snapshot_manifest() {
+            for (version, _) in snapshots {
```

Comment:
There's probably an obvious answer to this that I'm missing, but why do we need to check compatibility against all previous framework, instead of just the previous one?

Reply:
We don't have anything that strictly guarantees that snapshot files are compatible with each other (although the likelihood of this not holding is slim).

Reply:
Then we should check that version n is compatible with n-1, version n-1 is compatible with n-2, etc. Is checking the current framework against all previous versions good enough because compatibility is transitive?

Reply:
By checking the compatibility of the currently checked in framework with all previous snapshots, we are directly finding out whether the current framework is a valid upgrade of all historical snapshot options.

You can do the same thing by checking snapshot n Vs n-1 up the chain, but you need to assume transitivity for that, which does hold, but if it doesn't (because of a bug) you wouldn't find out if you were testing compatibility between snapshots, and you would with current formulation.

Reply:
> we are directly finding out whether the current framework is a valid upgrade of all historical snapshot options
Right I'm asking why we are checking this instead of checking whether Vn is valid upgrade of Vn-1, etc.

Reply:
That's what my second paragraph is meant to rationalize.  But here's another way I think about it:

The thing that's constantly changing, that we're trying to test is the framework that gets built into the binary, by testing it against every snapshot, we learn that the current framework binary can be used in an upgrade against any of the snapshots, whereas the test you propose only tells us that a sequence of hypothetical upgrades that ends with the framework in the binary is valid.

We can derive from this that we can upgrade from any of the snapshots to the current framework, but this requires assuming that compatibility is transitive (which it should be), but if this was not true because of a bug, we wouldn't learn about it -- the test would pass and a potential upgrade could fail.

Seeing as being able to upgrade to the framework in the binary is the property we are interested in, it seems better to me to test it directly.